### DISCUSSION TITLE
PSA: Test your textmate grammars in TextMate

### DISCUSSION THREAD

**Discussion by evoactivity** [github-linguist/linguist#6756] [General]
So with the latest release of linguist, we had included fixes for our glimmer-js and glimmer-ts grammars. We used NovaLightshow to test our grammar as it was capable of reproducing the issues we were seeing, and everything works perfectly there so we had hoped the same would be true for GitHub. It also works perfectly in VSCode and ShikiJS (and possibly other textmate derived syntax highlighting engines).

When we began testing the latest rollout of linguist everything seemed fine until we tested the simplest case our authoring format allows. Namely this

```glimmer-js
// if you're looking at this 3 months or more after this post hopefully you're
// seeing the correct highlighting. A screenshot is posted below for posterity 
const hello = 'Bonjour';

<template>
  <div>{{hello}} world</div>
</template>
```

<img width="383" alt="image" src="https://github.com/github-linguist/linguist/assets/83799/ceb73c20-8336-427a-b48f-3e48f9267ed3">

I was stumped on how to test this as it worked everywhere else except GitHub. That is until I decided to install TextMate and try our grammar there, what I saw was identical to how GitHub was highlighting (or not in this case).

<img width="326" alt="image" src="https://github.com/github-linguist/linguist/assets/83799/0364674c-a648-4258-b316-f19f3c68e7e5">

Now I had somewhere I could test I managed to fix the grammar

<img width="292" alt="image" src="https://github.com/github-linguist/linguist/assets/83799/591eb525-4ef8-4a5f-85e7-2018e16bcb45">

So if you grammar works everywhere but GitHub, and you're wondering how to most accurately test your grammar for GitHub support, I can definitely recommend using TextMate as it seems GitHub's highlighting engine matches TextMate's one-to-one, whereas other engines do not and might highlight things perfectly fine even if GitHub wont.

**Comment by RedCMD** [comment]
* [VSCode](https://github.com/microsoft/vscode)
  * [vscode-textmate](https://github.com/microsoft/vscode-textmate/tree/v9.2.0) v9.2.0
    * [vscode-oniguruma](https://github.com/microsoft/vscode-oniguruma/tree/v1.7.0) v1.7.0
      * [oniguruma](https://github.com/kkos/oniguruma/blob/v6.9.8/doc/RE) v6.9.8
* [TextMate](https://macromates.com/) 2.0
  * [textmate/Onigmo](https://github.com/textmate/Onigmo/blob/Onigmo-5.13.5/doc/RE) v5.13.5
    * [k-takata/Onigmo](https://github.com/k-takata/Onigmo)
      * [oniguruma](https://github.com/kkos/oniguruma)
* [GitHub](https://github.com/)
  * [Linguist](https://github.com/github-linguist)
  * PrettyLights
    * [PCRE](https://www.pcre.org/) v8.45
  * Pygments
* [NovaLightshow](https://github.com/Nixinova/NovaLightshow)
  * [starry-night](https://github.com/wooorm/starry-night) 
    * [vscode-textmate](https://github.com/microsoft/vscode-textmate)
      * [vscode-oniguruma](https://github.com/microsoft/vscode-oniguruma)
        * [oniguruma](https://github.com/kkos/oniguruma/blob/v6.9.8/doc/RE) v6.9.8
* [ShikiJS](https://github.com/shikijs/shiki)
  * [vscode-oniguruma](https://github.com/microsoft/vscode-oniguruma)
    * [oniguruma](https://github.com/kkos/oniguruma/blob/v6.9.8/doc/RE) v6.9.8

> seems GitHub's highlighting engine matches TextMate's one-to-one

it doesn't 
Github uses PCRE instead of TextMate's Onigmo
it also doesn't match TextMate's crazy `match`/`begin`/`end` shenanigans
doesn't allow comments within `captures`
and prob many other things

---

I assume this is your PR? https://github.com/lifeart/vsc-ember-syntax/pull/77
I found the reason for the problem
its due to a bug in TextMate 2.0, that Github then copied
VSCode did not copy the bug

in your grammar
`#main` is included in 2 places and applied twice in the same instance
both in the root level `patterns` and within the `injections`
`patterns` is checked first; for rules to apply
because `source.js` _was_ higher than `#main`, `source.js` had priority

the `injections` are then checked
the `L:` in the injection should have put the 2nd include of `#main` higher than the other 2 includes
but because `#main` has already been included in this instance
the 2nd reference was ignored
and thus the priority did not change

VSCode does not discard the 2nd `#main` (injection) and thus it then has priority over `source.js`

**Comment by DecimalTurn** [comment]
If someone like me who doesn't have a mac wants to test a grammar with TextMate, there is a way using a GitHub Action running on macOS like [this one](https://github.com/DecimalTurn/vscode-jsonc-syntax-highlighting/blob/main/.github/workflows/textmate-screenshot.yml) and you can get something like this :

![textmate](https://github.com/user-attachments/assets/f1d00140-30a3-4d98-84c4-9312d49f58fc)
